netket/models/mpdo.py

```diff
@@ -0,0 +1,234 @@
+# Copyright 2024 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.Mon
+
+from typing import Any, Optional
+
+from flax import linen as nn
+import jax
+from jax import numpy as jnp
+
+from jax.nn.initializers import normal
+
+from netket.hilbert import HomogeneousHilbert
+from netket.utils.types import NNInitFunc
+from netket.jax import dtype_complex
+
+default_kernel_init = normal(stddev=0.01)
+
+
+class MPDOPeriodic(nn.Module):
+    r"""
+    A Matrix Product Density Operator (MPDO) with periodic boundary conditions for a quantum mixed state of discrete
+    degrees of freedom. The purification is used.
+
+    The MPDO is defined as
+
+    .. math::
+        \rho(s_1,\dots s_N, s_1',\dots s_N') = \sum_{\alpha_1, \dots, \alpha_{N-1}} A^{\alpha_1}[s_1, s_1'] \dots A^{\alpha_{N-1}}[s_N, s_N'] \dots A^{\alpha_N}[s_1, s_1'],
+
+    for arbitrary local quantum numbers :math:`s_i`, where :math:`A^{\alpha_i}[s_i, s_i']` is a tensor
+    of dimensions (bdim, bdim), depending on the value of the local quantum number :math:`s_i` and
+    the bond index :math:`\alpha_{i}` connecting the adjacent tensors (the purification).
+
+    The periodic boundary conditions imply that there are connections between the first and the last tensors.
+    """
+
+    hilbert: HomogeneousHilbert
+    """Hilbert space on which the state is defined."""
+    bond_dim: int
+    """Bond dimension of the MPDO tensors."""
+    kraus_dim: int = 2
+    """Kraus dimension of the MPDO tensors."""
+    symperiod: Optional[bool] = None
+    """
+    Periodicity in the chain of MPDO tensors.
+    The chain of MPDO tensors is constructed as a sequence of identical
+    unit cells consisting of symperiod tensors. if None, symperiod equals the
+    number of physical degrees of freedom.
+    """
+    param_dtype: Any = jnp.float64
+    """complex or float, whether the variational parameters of the MPDO are real or complex."""
+    unroll: int = 1
+    """the number of scan iterations to unroll within a single iteration of a loop."""
+    kernel_init: NNInitFunc = default_kernel_init
```

Comment:
Is this a good choice? Should we have a MPS specific initialiser?

Reply:
For your information, ITensors uses [orthogonal init](https://github.com/ITensor/ITensors.jl/blob/9413240cc703a17abf6218951ba1171f3096802e/src/mps/mps.jl#L90) for `randomMPS` but I never compared it with Gaussian init

Reply:
I did the plus identity matrix trick for both MPS + MPDO which was there for MPSPeriodic. I found that to be better than without indeed. Would the orthogonal initializer be better and easy to implement too?